File: Data Structures & Algorithms/non-overlapping-intervals/submission-3.java
class Solution {
    public int eraseOverlapIntervals(int[][] intervals) {
        if (intervals.length == 0) return 0;
        int res = 0;
        Arrays.sort(intervals, (a, b) -> Integer.compare(a[0], b[0]));
        int preEnd = intervals[0][1];
        for (int i = 1; i < intervals.length; i++) {
            int[] cur = intervals[i];
            if (cur[0] < preEnd) {
                res += 1;
                preEnd = Math.min(preEnd, cur[1]);
            }
            else {
            preEnd = cur[1];

            }
            
        }
        return res;
    }
}

File: Data Structures & Algorithms/insert-new-interval/submission-3.py
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        start, end = newInterval
        res = []
        for cur_start, cur_end in intervals:
            if cur_end < start:
                res.append([cur_start, cur_end])
            elif cur_start <= start <= cur_end or cur_start <= end <= cur_end:
                start = min(cur_start, start)
                end = max(cur_end , end)
            elif cur_start > end:
                res.append([start, end])
                start, end = cur_start, cur_end
        res.append([start, end])
        return res
        # O(N), O(1 or N)